velox/dwio/parquet/thrift/ThriftStreamingTransport.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <thrift/transport/TVirtualTransport.h>
+#include "velox/dwio/common/BufferedInput.h"
+#include "velox/dwio/parquet/thrift/ThriftTransport.h"
+
+namespace facebook::velox::parquet::thrift {
+
+class ThriftStreamingTransport : public ThriftTransport {
+ public:
+  ThriftStreamingTransport(
+      std::shared_ptr<dwio::common::SeekableInputStream> inputStream,
+      const char* FOLLY_NULLABLE& bufferStart,
+      const char* FOLLY_NULLABLE& bufferEnd)
+      : ThriftTransport(),
+        inputStream_(inputStream),
+        bufferStart_(bufferStart),
+        bufferEnd_(bufferEnd) {
+    VELOX_CHECK_NOT_NULL(inputStream_);
+    VELOX_CHECK_NOT_NULL(bufferStart_);
+    VELOX_CHECK_NOT_NULL(bufferEnd_);
+  }
+
+  uint32_t read(uint8_t* outputBuf, uint32_t len) {
+    uint32_t remainLen = len;
+    while (remainLen > 0) {
+      uint32_t size = bufferEnd_ - bufferStart_;
+      auto copySize = std::min(size, remainLen);
+      memcpy(outputBuf, bufferStart_, copySize);
```

Comment:
We can allow `bufferStart_` to be `nullptr` so we don't need to call `SeekableInputStream::Next` ourselves outside this class.  In this case `bufferStart_` is also `nullptr` and `size` should be 0, and we just skip the copying, triggering the `bufferStart_ == bufferEnd_` branch.

Tracking `bufferSize_` instead of `bufferEnd_` will make the code a little simpler.

Reply:
OK we keep the `bufferEnd_` then